### PR TITLE
ref(ts): Convert stream group to typescript WOR-384

### PR DESCRIPTION
**Pull request by scttcper**
fixes JAVASCRIPT-2307

### CODE REVIEW

File: src/sentry/static/sentry/app/components/eventOrGroupHeader.tsx [hunk 1]
```diff
@@ -26,7 +26,7 @@ type Props = WithRouterProps<{orgId: string}> & {
   hideLevel?: boolean;
   query?: string;
   className?: string;
-} & DefaultProps;
+} & Partial<DefaultProps>;
```
Comment (matejminar) [inline src/sentry/static/sentry/app/components/eventOrGroupHeader.tsx:29]: Why this change?

File: src/sentry/static/sentry/app/components/stream/group.tsx [hunk 2]
```diff
@@ -420,7 +432,7 @@ const SecondaryCount = styled(({value, ...p}) => <Count {...p} value={value} />)
 const StyledMenuItem = styled(({to, children, ...p}) => (
   <MenuItem noAnchor>
     {to ? (
-      <Link to={to} target="_blank">
+      <Link to={to}>
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stream/group.tsx:435]: nice!..was it incorrect before?

File: src/sentry/static/sentry/app/components/stream/group.tsx [hunk 3]
```diff
@@ -183,8 +198,8 @@ const StreamGroup = createReactClass({
       };
 
       if (!!start && !!end) {
-        discoverQuery.start = start;
-        discoverQuery.end = end;
+        discoverQuery.start = start as string;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stream/group.tsx:201]:
```suggestion
        discoverQuery.start = String(start)
```

File: src/sentry/static/sentry/app/components/stream/group.tsx [hunk 4]
```diff
@@ -183,8 +198,8 @@ const StreamGroup = createReactClass({
...
+        discoverQuery.end = end as string;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stream/group.tsx:202]:
```suggestion
        discoverQuery.end = String(end);
```

File: src/sentry/static/sentry/app/components/stream/group.tsx [hunk 5]
```diff
@@ -61,37 +84,28 @@ const StreamGroup = createReactClass({
...
+  static defaultProps = defaultProps;
 
-  getInitialState() {
-    const data = GroupStore.get(this.props.id);
+  constructor(props: Props) {
+    super(props);
+    const data = GroupStore.get(props.id) as Group;
 
-    return {
+    this.state = {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stream/group.tsx:95]: should we move it out of the constructor? it's just a suggestion ;-)

File: src/sentry/static/sentry/app/components/eventOrGroupHeader.tsx (hunk 1 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/eventOrGroupHeader.tsx:29]: I am also having this issue in the file `components/eventOrGroupHeader.tsx`

File: src/sentry/static/sentry/app/components/stream/group.tsx (hunk 2 above)
Comment (scttcper) [reply src/sentry/static/sentry/app/components/stream/group.tsx:435]: said it was not a property? Will take a 2nd look at this

Comment (billyvg) [reply src/sentry/static/sentry/app/components/stream/group.tsx:435]: `to` means it's a react router link, so it wouldn't make sense to open them in a new tab (unless it was meant to be an external link)

File: src/sentry/static/sentry/app/components/stream/group.tsx (hunk 5 above)
Comment (scttcper) [reply src/sentry/static/sentry/app/components/stream/group.tsx:95]: I didn't like how it looked with GroupStore needing to be called multiple times

File: src/sentry/static/sentry/app/components/stream/group.tsx (hunk 2 above)
Comment (davidenwang) [reply src/sentry/static/sentry/app/components/stream/group.tsx:435]: I think @taylangocmen intended for this to open in a new tab not sure how to properly do that with typescript tho

Comment (billyvg) [reply src/sentry/static/sentry/app/components/stream/group.tsx:435]: Use `<ExternalLink />`

File: src/sentry/static/sentry/app/components/stream/group.tsx [hunk 6]
```diff
@@ -251,14 +262,14 @@ const StreamGroup = createReactClass({
...
-          <EventOrGroupHeader data={data} query={query} />
+          <EventOrGroupHeader includeLink data={data} query={query} size="normal" />
           <EventOrGroupExtraDetails data={data} />
         </GroupSummary>
         {hasGuideAnchor && <GuideAnchor target="issue_stream" />}
         {withChart && (
           <Box width={160} mx={2} className="hidden-xs hidden-sm">
             <GroupChart
-              statsPeriod={statsPeriod}
+              statsPeriod={statsPeriod!}
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stream/group.tsx:272]: Are you sure about it? I am always trying to avoid the usage of `something!`

File: src/sentry/static/sentry/app/components/stream/group.tsx [hunk 7]
```diff
@@ -100,57 +110,58 @@ const StreamGroup = createReactClass({
...
       return true;
     }
     return false;
-  },
+  }
 
   componentWillUnmount() {
-    clearTimeout(this.hoverWait);
-  },
+    callIfFunction(this.listener);
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/stream/group.tsx:126]: 👏

File: src/sentry/static/sentry/app/components/links/link.tsx [hunk 8]
```diff
@@ -15,7 +15,7 @@ type Props = {
   to: string | ToLocationFunction | LocationDescriptor;
   // Styles applied to the component's root
   className?: string;
-} & Omit<AnchorProps, 'href' | 'target'>;
+} & Omit<AnchorProps, 'href'>;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/links/link.tsx:18]: I wouldn't remove the target from the Omit. The idea was that if need to open a link in a new tab you should use the component `<ExternalLink />`

File: src/sentry/static/sentry/app/components/stream/group.tsx (hunk 5 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/stream/group.tsx:91]: I see...

I believe it's fine if you do something like:

```
  state: State = this.getInitialState();

  getInitialState() {
    const {id, useFilteredStats} = this.props;

    const data = GroupStore.get(id) as Group;

    return {
      data: {
        ...data,
        filtered: useFilteredStats ? data.filtered : undefined,
      },
    };
  }
```

File: src/sentry/static/sentry/app/components/links/link.tsx (hunk 8 above)
Comment (scttcper) [reply src/sentry/static/sentry/app/components/links/link.tsx:18]: Talked with @billyvg about this, this is one of the few times we're opening a new page within our own application. External link was made for links to outside websites and takes a href and we need to use the RouterLink's to for this link to another place within our app.

I also asked the team if they meant for it to open in another tab, and it seems pretty much everyone wanted it to keep opening in a new window.

Comment (billyvg) [reply src/sentry/static/sentry/app/components/links/link.tsx:18]: Yeah, I think it's better to allow an exception in this case because 1) it opens a new tab to our app, 2) the URL would require stringify-ing the query string. But because it's an exception, would it be better to `@ts-ignore` where we use it and explain why?

Comment (scttcper) [reply src/sentry/static/sentry/app/components/links/link.tsx]: switched to ts-ignore for this case

File: src/sentry/static/sentry/app/components/stream/group.tsx (hunk 6 above)
Comment (scttcper) [reply src/sentry/static/sentry/app/components/stream/group.tsx:275]: statsPeriod has a default prop, but is currently somewhat broken because of the higher order functions on export. Had to cast default props to partial...

**Review by davidenwang — APPROVED** [review]

**Review by billyvg — APPROVED** [review]

**Review by priscilawebdev — APPROVED** [review]
Good Job @scttcper  👏

**Review by matejminar — APPROVED** [review]
